trace-utils/src/msgpack_decoder/v04/decoder/span.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use super::{
+    read_map_strs, read_meta_struct, read_metrics, read_string, read_string_ref,
+    span_link::read_span_links,
+};
+use crate::msgpack_decoder::v04::error::DecodeError;
+use crate::msgpack_decoder::v04::number::read_number;
+use datadog_trace_protobuf::pb::Span;
+use std::str::FromStr;
+
+/// Decodes a slice of bytes into a `Span` object.
+///
+/// # Arguments
+///
+/// * `buf` - A mutable reference to a slice of bytes containing the encoded data.
+///
+/// # Returns
+///
+/// * `Ok(Span)` - A decoded `Span` object if successful.
+/// * `Err(DecodeError)` - An error if the decoding process fails.
+///
+/// # Errors
+///
+/// This function will return an error if:
+/// - The map length cannot be read.
+/// - Any key or value cannot be decoded.
+#[inline]
+pub(crate) fn decode_span(buf: &mut &[u8]) -> Result<Span, DecodeError> {
+    let mut span = Span::default();
+    let span_size = rmp::decode::read_map_len(buf).map_err(|_| {
+        DecodeError::InvalidFormat("Unable to get map len for span size".to_owned())
+    })?;
+
+    for _ in 0..span_size {
+        fill_span(&mut span, buf)?;
+    }
+    Ok(span)
+}
+
+#[derive(Debug, PartialEq)]
+enum SpanKey {
+    Service,
+    Name,
+    Resource,
+    TraceId,
+    SpanId,
+    ParentId,
+    Start,
+    Duration,
+    Error,
+    Meta,
+    Metrics,
+    Type,
+    MetaStruct,
+    SpanLinks,
+}
+
+impl FromStr for SpanKey {
+    type Err = DecodeError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "service" => Ok(SpanKey::Service),
+            "name" => Ok(SpanKey::Name),
+            "resource" => Ok(SpanKey::Resource),
+            "trace_id" => Ok(SpanKey::TraceId),
+            "span_id" => Ok(SpanKey::SpanId),
+            "parent_id" => Ok(SpanKey::ParentId),
+            "start" => Ok(SpanKey::Start),
+            "duration" => Ok(SpanKey::Duration),
+            "error" => Ok(SpanKey::Error),
+            "meta" => Ok(SpanKey::Meta),
+            "metrics" => Ok(SpanKey::Metrics),
+            "type" => Ok(SpanKey::Type),
+            "meta_struct" => Ok(SpanKey::MetaStruct),
+            "span_links" => Ok(SpanKey::SpanLinks),
+            _ => Err(DecodeError::InvalidFormat(
+                format!("Invalid span key: {}", s).to_owned(),
+            )),
+        }
+    }
+}
+
+fn fill_span(span: &mut Span, buf: &mut &[u8]) -> Result<(), DecodeError> {
+    // field's key won't be held so no need to copy it in a buffer.
+    let key = read_string(buf)?;
+    let key = key.parse::<SpanKey>()?;
```

Comment:
This code actually needs to use `read_string_ref` since we don't want to materialize a `String` just to throw it away. It has a big performance impact.